cpp/dolfinx/io/checkpointing.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright (C) 2024 Abdullah Mujahid
+//
+// This file is part of DOLFINX (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#ifdef HAS_ADIOS2
+
+#include "checkpointing.h"
+#include "ADIOS2_utils.h"
+#include <adios2.h>
+#include <basix/finite-element.h>
+#include <dolfinx/mesh/Mesh.h>
+#include <mpi.h>
+
+using namespace dolfinx;
+using namespace dolfinx::io;
+
+/// @file checkpointing.h
+/// @brief ADIOS2 based checkpointing
+namespace
+{
+std::map<basix::element::lagrange_variant, std::string> lagrange_variants{
+    {basix::element::lagrange_variant::unset, "unset"},
+    {basix::element::lagrange_variant::equispaced, "equispaced"},
+    {basix::element::lagrange_variant::gll_warped, "gll_warped"},
+};
+
+template <std::floating_point T>
+void _write(ADIOS2Engine& adios2engine,
+            std::shared_ptr<dolfinx::mesh::Mesh<T>> mesh)
+{
+
+  auto io = adios2engine.io();
+  auto writer = adios2engine.engine();
+
+  const dolfinx::mesh::Geometry<T>& geometry = mesh->geometry();
+  auto topology = mesh->topology();
+
+  std::int16_t mesh_dim = geometry.dim();
+
+  auto imap = mesh->geometry().index_map();
+  std::uint64_t num_nodes_global = imap->size_global();
+  std::uint32_t num_nodes_local = imap->size_local();
+  std::uint64_t offset = imap->local_range()[0];
+
+  const std::shared_ptr<const dolfinx::common::IndexMap> topo_imap
+      = topology->index_map(mesh_dim);
+  std::uint64_t num_cells_global = topo_imap->size_global();
+  std::uint32_t num_cells_local = topo_imap->size_local();
+  std::uint64_t cell_offset = topo_imap->local_range()[0];
+
+  auto cmap = mesh->geometry().cmap();
+  auto geom_layout = cmap.create_dof_layout();
+  std::uint32_t num_dofs_per_cell
+      = geom_layout.num_entity_closure_dofs(mesh_dim);
+
+  const std::vector<int64_t> input_global_indices
+      = geometry.input_global_indices();
+  const std::span<const int64_t> input_global_indices_span(
+      input_global_indices.begin(), num_nodes_local);
+  const std::span<const T> mesh_x = geometry.x();
+
+  auto connectivity = topology->connectivity(mesh_dim, 0);
+  auto topology_array = connectivity->array();
+  auto topology_offsets = connectivity->offsets();
+
+  const std::span<const int32_t> topology_array_span(
+      topology_array.begin(), topology_offsets[num_cells_local]);
+  std::vector<std::int64_t> topology_array_global(
+      topology_offsets[num_cells_local]);
+
+  std::iota(topology_array_global.begin(), topology_array_global.end(), 0);
+
+  imap->local_to_global(topology_array_span, topology_array_global);
+
+  for (std::size_t i = 0; i < num_cells_local + 1; ++i)
+    topology_offsets[i] += cell_offset * num_dofs_per_cell;
+
+  const std::span<const int32_t> topology_offsets_span(topology_offsets.begin(),
+                                                       num_cells_local + 1);
+
+  io->DefineAttribute<std::string>("name", mesh->name);
```

Comment:
Need to add metadata e.g. DOLFINx version, DOLFINx git hash etc.

Reply:
How to do this?

I tried

```
#include <dolfinx/common/version.h.in>

  std::cout << "version" << DOLFINX_VERSION_STRING << std::endl;

  std::cout << "DOLFINX_GIT_COMMIT_HASH" << DOLFINX_VERSION_GIT << std::endl;
```

which prints

```
version@DOLFINX_VERSION@
DOLFINX_GIT_COMMIT_HASH@GIT_COMMIT_HASH@
version@DOLFINX_VERSION@
DOLFINX_GIT_COMMIT_HASH@GIT_COMMIT_HASH@

```

Reply:
These are replaced with the correct strings only after `cmake install` and the correct include is then `#include <dolfinx/common/version.h>`